Gracefully bail out on extension method rewrite within conditional ac…

mavasani opened the pull request:
…cess expression

**Fixes #2593**

**User scenario:** User types a conditional access expression with a subsequent extension method invocation. For example, "a?.Method().ExtensionMethod()". Attempting to perform any simplification (cast simplification, simplify name) or refactorings (inline temporary) causes a exceptions in the simplification engine and all these IDE services are disabled for the project.

**Issue description:** We are not handling conditional access expressions, and hence member access or invocations with null "left of dot" nodes, in our expander. Additionally, performing extension method expansion (to a static method invocation with receiver replaced as the first argument to invocation) is non-trivial.

**Fix description:** This change makes a defensive fix by:

1. Guarding against null reference exceptions
2. Bailing out on extension method rewrite within conditional access expressions.

Long term fix would be to implement this feature in expand/reduce engine, but this should happen post 1.0 milestone. A separate issue will be opened to track this work.

**Testing:** Added regression tests + existing tests.

/cc @srivatsn @ManishJayaswal This is for 1.0 stable milestone.

mavasani commented:
@shyamnamboodiripad @jmarolf @tmeschter @heejaechang @JohnHamby @balajikris can you please review?

tmeschter commented:
@mavasani Is it my understanding from the unit tests that Inline Temporary does work with extension methods in conditional access expressions?

Also, could you add a skipped unit test demonstrating what doesn't work, linked to the issue to track the future feature work?

mavasani commented:
Yes Tom, that's correct. I'll add a skipped test and open work item for the same.

mavasani commented:
To clarify, we are skipping the step in the expander that converts an extension method invocation to a regular static method invocation with the receiver as the first argument. Any simplifications relying on this form of invocation for extension methods will not work, until we fix this in future. I'll bash around this is a bit more to verify simplifier doesn't blow up with the current workaround.

balajikris commented on src/Workspaces/CSharp/Portable/Simplification/CSharpSimplificationService.Expander.cs:
```diff
@@ -930,6 +930,13 @@ public override SyntaxNode VisitInvocationExpression(InvocationExpressionSyntax
+                var originalMemberAccess = (MemberAccessExpressionSyntax)originalNode.Expression;
+                if (originalMemberAccess.GetParentConditionalAccessExpression() != null)
+                {
+                    // Bail out on extension method invocations in conditional access expression.
```
Nit: If possible, tag the GH issue tracking this Todo work here, so that when someone looks at it in the future, they understand this is simply a workaround and not an intentional design.

balajikris commented:
:+1:  Looks good to me, pending adding skipped tests for simplification scenarios that do not work and a GH workitem tracking it.

mavasani commented:
@tmeschter @balajikris I have addressed your feedback and opened https://github.com/dotnet/roslyn/issues/3260 and added skipped unit tests (I have also added the corresponding tests with an equivalent member access instead of conditional access expression, which pass fine and we would want the conditional access ones to also behave similarly)

balajikris commented:
thanks, LGTM.

tmeschter commented:
LGTM

srivatsn commented:
LGTM

MattGertz commented:
Approved 6/4 by ML Shiproom.
